### ISSUE TITLE
The end of the add-on development on Kodi 18.x ⏰ it is time to move on

### ISSUE THREAD

**Issue by CastagnaIT** [CastagnaIT/plugin.video.netflix#975]
### The add-on version 1.12.0 will be the last add-on _feature_ update for Kodi 18.x

**What means, what you should expect now on Kodi 18.x ?**
- Only critical updates will be applied (so bug fixing or minor changes to restore add-on functionality)
- Critical updates _that require too much work_ to the source code will _not_ be done directly from maintainer, but a backport (from master branch) made by other developers can be accepted
- The backport of new features or changes in the existing features developed in the add-on for Kodi 19 version, could be accepted only if they are maintained by other developers, the maintainer will not provide support to fix any side effects or backporting errors

**Why this choice by the maintainer**
Maintaining the add-on on two platforms (Kodi 18 and Kodi 19) and support all users on both platforms, takes a lot of time of work, in the meantime, development of a new platform is about to start (Kodi 20) which will add a further weighting to current maintenance.
To lighten the whole management it is necessary to clean up the code and support only the new generations of Kodi based on Python 3.

The _Release Candidate 1_ of Kodi 19 will be released soon, this release will almost definitively establish what will be the final version of Kodi 19. There are many different systems using Kodi, some are still late in getting ready for Kodi 19, other systems will take a long time to achieve Kodi 19 (like on RPI), but who has the chance, start try switch to Kodi 19 when will be released.

**Comment by dagwieers** [comment]
I think it is too soon to drop Kodi v18 development. Even today less than 0.7% of Kodi users is (able to) run Kodi v19
(based on the latest release of ISH: http://mirrors.kodi.tv/stats?downloadstats&filter=script.module.inputstreamhelper).

I expect that even one year after the release of Kodi v19 a majority will still be running Kodi v18 out of necessity.

What are the issues to maintain the add-on for both Kodi v18 and Kodi v19? Except for the addon.xml, we have few issues maintaining both in a single codebase.

**Comment by CastagnaIT** [comment]
i undestand your POV

But i'm no longer willing to maintain a double source code for two platforms in the same modules
sometimes it involves a hard work that takes a long time to test and solve problems/requests of all users,
this because the source code is no longer linear and remembering all the particular cases every time is difficult and this causes side effects.

As i wrote before these are my conditions, only critical bugs for kodi 18.
The basic features are already all there on Kodi 18, but those who want to be more up to date or new features in the future will have to switch to Kodi 19.

This may not suit everyone, but i need to simplify the code and make things simpler
or else i will abandon the project because it is taking up too many personal resources.

**Comment by dagwieers** [comment]
Of course, as a volunteer you decide how to use your time best, no arguing about that ! I just wonder if there is no middle ground.

If tomorrow Netflix decides to redo their authentication completely, almost all Kodi users will be without a working Netflix. Some may be locked out because their platform may not/never have Kodi v19.

**Comment by vascobraga41** [comment]
"As i wrote before these are my conditions, only critical bugs for kodi 18."

I think you skipped this part CastagnaIT wrote.

**Comment by dagwieers** [comment]
@vascobraga41 A new authentication framework is not a bugfix, it is new development that may require backporting from the Kodi v19 branch.

> Critical updates that require too much work to the source code unfurnately will not be done

So the timing for making 2 branches is what is leading up to all of this. And that is what I think is too soon. I understand Netflix is not your typical Kodi add-on (as it touches much more Kodi specific calls and quirks), but avoiding two branches is paramount at this stage IMO.

**Comment by vascobraga41** [comment]
You're right and I hope it doesn't get to that. But there's only so much one person can do and if we get to that maybe someone can help with the backport from Matrix. And things are moving fast so I believe there will be Kodi Matrix available for most devices soon.

**Comment by dagwieers** [comment]
@vascobraga41 Again, almost 12% of our Kodi users are still running Kodi v17 (Krypton) 2 years after Kodi v18 was released (jan 2019). So I want to believe the majority will move to Kodi v19 asap, but realistically I think we will have the majority of Kodi users on Kodi v18 for the next 6 to 12 months. Which is something to consider before creating two separate branches.

The other problem is indeed that the Netflix add-on is developed by a single person mostly. I don't know how we can fix that easily, the codebase has evolved a lot. Certainly not something you can easily dive into.

PS I am not criticising @CastagnaIT's work, not at all. I have a lot of respect of the work he is doing, that should be made very clear!

**Comment by CastagnaIT** [comment]
@dagwieers i understood what you want to say and i agree to a certain degree

this project is big and here you have to take in account of manage on same modules the platform variables for
Kodi 18, Kodi 19 and near future next Kodi 20, and plus every operative systems due to multi-platform system,
this cause confusion in the code and difficulty in understanding everything simply way,
therefore external help from new developers is also difficult due to the fragmentation of the code.

from other side there are problems that you can easily fix and end in fast way, but others require **days** of testing on different Kodi versions and sometimes run the same tests on different operative systems because unexpected behaviour happens

moreover, being tied to having to keep a continuous backward compatibility with Python 2 (of Kodi 18) is a limitation that undermines possible better development solutions in the project

As the only developer, i need to improve all this things for conciling better also with my life,
so: Simplify the code -> lighten the maintenance workload --> lighten the future development
this allows me to spend less personal time over the time to manage everything (that include GitHub Issues requests of users)
and the release of Kodi 19 is the best time to do this

the transition of users from kodi 18 to kodi 19 will be very long, and will also can take a year or more!

this is not a professional work and is not supported by anyone
so the users will also have to adapt, and can help by maintains the Kodi 18 version by backporting and supporting related GitHub Issue requests

**Comment by wagnerch** [comment]
I'm happy to help with backporting bugs/features to the v18 code base, as long as I am still using v18.  I'm just not aware of what the needs are, my personal point of view is the addon is currently fairly feature complete for what I use it for and I would be pretty content with bug fixes only.

At some point, I will move on to v19.  The question is when, because v19 is pretty iffy still for Linux-based platforms due to V4L2 seek issues.  Especially, RPi has no hardware decoding of H.264 due to seeking issues with V4L2, so that will keep me off v19 until it is resolved.  So to @dagwieers point, I will probably be on v18 quite a bit longer and may not upgrade until v20 -- really depends on where V4L2 issues are at.

I don't know if you looked at perhaps designing abstract classes to keep the main add-on code simple and behind the abstract classes you could have a factory class that spits out a platform specific class (Linux, Windows, Android, etc).  The other part is perhaps a factory class that spits out Kodi classes.  But that may help with some of the complexity.  The worst part is import & class variations between Python 2 & 3, that stuff is pretty obnoxious, not sure there is a whole lot that can be done there.

**Comment by wagnerch** [comment]
Just to add to my earlier comment, I don't see a problem with 2 branches.  But I think if we could split out v18 & v19 Kodi logic into abstraction classes then a lot of the backporting could be simple cherry-pick's, and implementating new methods in the abstract class for v18.   There is going to be some things v19 can do and v18 won't, then the question is how we handle those scenarios with the abstract class, it might be a matter of "do nothing" or manage something via DB/memory/cache.  Really would have to design the code to anticipate the possibility that the feature doesn't exist.  Maybe the abstraction classes could have a method "hasFeature(<FEATURE NAME>)" and that is used in the main add-on code, etc.

**Comment by CastagnaIT** [comment]
@wagnerch it is good if you want take care of backports for v18 addon version,
keep in mind that if you backport new features or future feature changes, you will also have to do support for any problems, personally i prefeer to avoid, but if you want take the task to support this too then no problem

i will not create new abstract classes, weighs down the entire infrastructure,
in the master branch i will delete all python 2 code, "futures" modules correlated and all code dedicated to Kodi 18 and the workarounds

**Comment by dm2912** [comment]
Thing to bear in mind in this is that there are just "kodi" users. 

They have a fairly easy upgrade path to 19 once they have ensured that all their add-ons are supported. But there is a wider community using CoreElec, OpenElec, LibreElec etc etc 

For them an update to a v19 base is unlikely to happen anytime soon. And for some devices, it may not happen at all.

**Comment by donkeykin** [comment]
I am using the plugin on Libreelec with Kodi 18.9. The new update is requesting python 3. I understand you want to drop support for Kodi 18.x, so should I follow a new fork of this project instead? Is Wagnerich handling that?

All Kodi 18.x users will not be able to install any new updates for this plugin, I guess. So how can we maintain functionality of the plugin moving forward? For now the plugin continues to function, but that could change.

Many thanks.

**Comment by notoco** [comment]
The new update for Leia v1.12.2 does not require python3 - you had to install the Matrix repository, not Leia.

> All Kodi 18.x users will not be able to install any new updates for this plugin, I guess. So how can we maintain functionality of the plugin moving forward?

By installing another STABLE version of kodi - k19.

**Comment by dagwieers** [comment]
I reported this to @CastagnaIT privately, but the TVADDONS.CO repository has included the latest Netflix release in the Leia repository, causing this issue for Leia users that have this repository enabled.

**Comment by notoco** [comment]
As always, TVAddons sucks ... Who normal uses it?

**Comment by CastagnaIT** [comment]
the people who create these repositories should ensure that they are maintained
i will not do maintenance for others...

**Comment by dagwieers** [comment]
This is by the way not a unique problem, I got a similar Kodi Matrix add-on (Retro BIOS Tool) pushed via **Zach Morris Add-ons** to Kodi Leia which is indirectly included in LibreELEC.

**Comment by dm2912** [comment]
> @donkeykin I reported this to @CastagnaIT privately, but the TVADDONS.CO repository has included the latest Netflix release in the Leia repository, causing this issue for Leia users that have their repository enabled.
> 
> If this is your problem, please report it to TVADDONS.CO.

Nobody who has any concern for their Kodi install should have tvaddons repo installed. Its toxic. It has frequently delivered malware, it forces addons to install that arent needed like Indigo, and it even goes as far as totally disabling addons where Adam (tva) has a personal issue with the dev.. (For instance, he fell out with the devs of the Crew so indigo now forcibly removes that addon from your installs).

In short, Tvaddons is toxic malware and should be avoided.. and dont even get me started on Adam's alterego (kodibae)

**Comment by dm2912** [comment]
So is there a fork of Netflix for Kodi 18?

I really do think there is a tremendous overestimation of just how quickly people are going to move to Kodi 19. For me, Coreelec isnt even close to v10 final yet so I won't be updating for some time. 

I only use a tiny handful of addons, but until they all support 19 I wont be moving, and its very possible if some like Open VPN manager don't update then I will stay on Kodi 18 for a good year yet. 

Deserting Kodi 18 before 19 is even final, and a long time before LibreElec and OpenElec are final kinda sucks..

**Comment by CastagnaIT** [comment]
@dm2912 in the readme there are the links for Kodi 18 and Kodi 19 repo version
you have only to choose it...

**Comment by dagwieers** [comment]
As I was suspecting, LibreELEC is going to hold back supporting Kodi v19 until they can offer a good experience to its users. https://libreelec.tv/2021/02/upcoming-changes/

**Comment by dm2912** [comment]
> As I was suspecting, LibreELEC is going to hold back supporting Kodi v19 until they can offer a good experience to its users. https://libreelec.tv/2021/02/upcoming-changes/

not a surprise, they also talk in that post about the amlogic issues which i think will see Coreelec supporting v9/18 for the foreseeable as well.  Personally as a coreelec user I see myself on v9/18 for a good year yet. 

hopefully @wagnerch is backporting fixes etc to kodi 18, would be good to have some clarification on this.

**Comment by CastagnaIT** [comment]
wagnerch proposed to backport future significant changes,
but whoever does this, will have the responsibility for supporting any backport problems and follow new opened related GH issues

I was no longer contacted by wagnerch, and since he hasn't replied here anymore, suggests that he won't support this add-on, at least not in this moment

the future planned big changes for the addon includes:
- Changes to addon services, to join all http servers, with a single multithreaded server and a single websession
- A first test to drop Requests module and use the HTTP/2 web communication (recently introduced by Netflix servers)

these changes will not be easy to backport to version 18, especially as regards the future HTTP/2,
HTTP/2 will not be a required step in the short term, but could be an issue if Netflix website drops HTTP1.1 definitively

**Comment by wagnerch** [comment]
> wagnerch proposed to backport future significant changes,
> but whoever does this, will have the responsibility for supporting any backport problems and follow new opened related GH issues
> 
> I was no longer contacted by wagnerch, and since he hasn't replied here anymore, suggests that he won't support this add-on, at least not in this moment

I don't really see a need for me to step in and help with v18 backports, I think I am in the same position as you.  As I mentioned in my comment, I feel like the v18 add-on is fairly feature complete with Netflix.  And since you are already committed to supporting bug fixes for now, it seems like it is covered.

I really don't think it will be long for me on v18, Kodi 19 & kernel changes have come along for the RPi4, and it is essentially ready.  But I will probably wait until 19.1 or 19.2, so it has a chance to stabilize.  Vero4K is another story, I haven't seen any builds for their device in 7 months, and it seems like some Amlogic devices will not make the Kodi 19 cut based on what CoreELEC has put out.

 
>     * Changes to addon services, to join all http servers, with a single multithreaded server and a single websession

I think this is a good idea.  The add-on services seem a bit complicated.  Perhaps, you can reduce the need on calling caching & IPC over TCP loopback, I feel like this adds some latency to the experience but have no data on it.

>     * A first test to drop Requests module and use the HTTP/2 web communication (recently introduced by Netflix servers)

I would hope that there is a HTTP 2.0 module for Python 2.7 that is requests compatible, or is that not the case?  But as you pointed out it might not be a problem for the life left in Kodi 18.

**Comment by CastagnaIT** [comment]
> I think this is a good idea. The add-on services seem a bit complicated. Perhaps, you can reduce the need on calling caching & IPC over TCP loopback, I feel like this adds some latency to the experience but have no data on it.

currently 90% of data processing has been moved within the service itself, then IPC is little used,
except two things (1 read below) and when you run "sync kodi library" manually from addon settings, but this is another story.

The IPC can be used from frontend instance also to read-write addon DB data (like settings or library) and use cache methods.
This due to the sqlite and his bad management of concurrency/data sharing with multiple read/write access,
because Kodi can execute an add-on (frontend) instance many times at the same time (so different python executions) and this can broke the sqlite db's especially on slow storages.
To solve this the addon service handle the db's access (possible single connection), and the frontend will access to the db's  through IPC, managing the DB's directly from the frontend will be faster (we are talking about of milliseconds) and yes perhaps will make the code more soft and easy, but i suspect it will lead to regressions

it is not clear what you mean by "reducing need on calling caching",
if you remove caching, all low-end devices will become very slow to load lists,
the use of caching has two purposes, avoid CPU data re-elaboration, avoid make every time new http requests when you browse the add-on lists. This improve a lot the low-end devices (and also some fast devices), there is no big latence on powerful devices.

However the merging of the http servers in to one, should in theory solve some particular cases like some internal IPC calls (service to service), i will need time to figure out the best strategy, i haven't done anything yet...

> I would hope that there is a HTTP 2.0 module for Python 2.7 that is requests compatible, or is that not the case? But as you pointed out it might not be a problem for the life left in Kodi 18.

In the past i did a search for HTTP/2 for python 2 seem there is only a module called hyper but is outdated, could still work i do not know...

**Comment by Paderman** [comment]
I'm very sorry for this question, but what is the workable Netflix plugin for Kodi 19?
This plugin is not compatible...

I understand that not more versions can supported for a long time.

Thank you :smiley:
